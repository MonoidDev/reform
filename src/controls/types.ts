import { Either } from '../types/Either';
import { ErrorMessage } from '../types/ErrorMessage';
import type { FormControl } from './FormControl'
import type { StructFormControl } from './StructFormControl'

export type AnyFormControl = FormControl<any, any, any, any> | StructFormControl<any>;

export type FormResult<O> = Either<ErrorMessage | undefined, O>;

export type Unsubscribe = () => void;
