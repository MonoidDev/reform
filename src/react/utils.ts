import { useState } from 'react';

export const useForceUpdate = () => {
  const [, s] = useState({});
  return () => s({});
}
