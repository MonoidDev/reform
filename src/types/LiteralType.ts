import { makeLeft, makeRight } from './Either';
import { ErrorMessage } from './ErrorMessage';
import { makeResolver, Resolver } from './Resolver';

export type LiteralValue = string | number | boolean | null | undefined;

export interface TLiteral<T extends LiteralValue>
  extends Resolver<unknown, T, ErrorMessage, 'literal'> {
  value: T;
}

export function literal<T extends LiteralValue>(value: T): TLiteral<T> {
  return {
    ...makeResolver<unknown, T, ErrorMessage, 'literal'>('literal', (input) => {
      if (input === value) {
        return makeRight(input as T);
      } else {
        return makeLeft({
          message: `Expect ${value}, got ${input}. `,
        });
      }
    }),
    value,
  };
}
