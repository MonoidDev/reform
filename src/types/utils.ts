export const isObject = (input: unknown): input is {} => {
  return (
    (typeof input === 'object' || typeof input === 'function') && input !== null
  );
};
