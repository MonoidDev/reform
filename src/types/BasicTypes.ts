import { isLeft, makeLeft, makeRight } from './Either';
import { ErrorMessage } from './ErrorMessage';
import { AnyResolver, OutputOf } from './infer';
import {
  ErrorContext,
  makeErrorContext,
  makeResolver,
  Resolver,
} from './Resolver';
import { isObject } from './utils';

export interface TAny extends Resolver<any, any, ErrorMessage, 'any'> {}

export function any(): TAny {
  return {
    ...makeResolver<any, any, ErrorMessage, 'any'>('any', (input) =>
      makeRight(input)
    ),
  };
}

export interface TString
  extends Resolver<unknown, string, ErrorMessage, 'string'> {}

export function string(): TString {
  return {
    ...makeResolver<unknown, string, ErrorMessage, 'string'>(
      'string',
      (input) => {
        if (typeof input === 'string') {
          return makeRight(input);
        } else {
          return makeLeft({
            message: `Expect a string, got ${typeof input}. `,
          });
        }
      }
    ),
  };
}

export interface TNumber
  extends Resolver<unknown, number, ErrorMessage, 'number'> {}

export function number(): TNumber {
  return {
    ...makeResolver<unknown, number, ErrorMessage, 'number'>(
      'number',
      (input) => {
        if (typeof input === 'number') {
          return makeRight(input);
        } else {
          return makeLeft({
            message: `Expect a number, got ${typeof input}. `,
          });
        }
      }
    ),
  };
}

export interface TBoolean
  extends Resolver<unknown, boolean, ErrorMessage, 'boolean'> {}

export function boolean(): TBoolean {
  return {
    ...makeResolver<unknown, boolean, ErrorMessage, 'boolean'>(
      'boolean',
      (input) => {
        if (typeof input === 'boolean') {
          return makeRight(input);
        } else {
          return makeLeft({
            message: `Expect a boolean, got ${typeof input}. `,
          });
        }
      }
    ),
  };
}

export interface TNull extends Resolver<unknown, null, ErrorMessage, 'null'> {}

export function nullType(): TNull {
  return {
    ...makeResolver<unknown, null, ErrorMessage, 'null'>('null', (input) => {
      if (input === null) {
        return makeRight(input);
      } else {
        return makeLeft({
          message: `Expect a null, got ${typeof input}. `,
        });
      }
    }),
  };
}

export interface TUndefined
  extends Resolver<unknown, undefined, ErrorMessage, 'undefined'> {}

export function undefinedType(): TUndefined {
  return {
    ...makeResolver<unknown, undefined, ErrorMessage, 'undefined'>(
      'undefined',
      (input) => {
        if (input === undefined) {
          return makeRight(input);
        } else {
          return makeLeft({
            message: `Expect an undefined, got ${typeof input}. `,
          });
        }
      }
    ),
  };
}

export interface TObject
  extends Resolver<unknown, {}, ErrorMessage, 'object'> {}

export function object(): TObject {
  return {
    ...makeResolver<unknown, {}, ErrorMessage, 'object'>('object', (input) => {
      if (isObject(input)) {
        return makeRight(input);
      } else {
        return makeLeft({
          message: `Expect an object, got ${typeof input}. `,
        });
      }
    }),
  };
}

export interface TArray<T extends AnyResolver>
  extends Resolver<unknown, OutputOf<T>[], ErrorMessage, 'array'> {}

export function array<T extends AnyResolver>(t: T): TArray<T> {
  return {
    ...makeResolver<unknown, OutputOf<T>[], ErrorMessage, 'array'>('array'),
    resolve(input, contexts = [makeErrorContext()]) {
      if (!Array.isArray(input)) {
        contexts[contexts.length - 1].error = {
          message: `Expect an array, got ${typeof input}`,
        };
        return makeLeft(contexts);
      }

      let value: any = [];

      const newErrorContexts: ErrorContext[] = []; 
      for (let i = 0; i < input.length; i++) {
        const nextContext: ErrorContext = {
          inputPath: [...contexts[contexts.length - 1].inputPath, i],
          resolverPath: [...contexts[contexts.length - 1].resolverPath],
          error: undefined,
        };

        const result = t.resolve(input[i], [nextContext]);

        if (isLeft(result)) {
          newErrorContexts.push(...result.left);
        } else {
          value.push(result.right);
        }
      }

      if (newErrorContexts.length) {
        contexts.pop(); // Remove last context because there're no errors
        return makeLeft([
          ...contexts,
          ...newErrorContexts,
        ]);
      }

      return makeRight(value);
    }
  };
}
