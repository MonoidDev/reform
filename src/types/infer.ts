import { TAny } from "./BasicTypes"
import { Resolver } from "./Resolver"

export type AnyResolver = Resolver<any, any, any, string>;

export type ResolverMap = {
  [K in string]: AnyResolver;
}

export type InputOfResolverMap<RM extends ResolverMap> = {
  [K in keyof RM]: InputOf<RM[K]>;
}

export type OutputOfResolverMap<RM extends ResolverMap> = {
  [K in keyof RM]: OutputOf<RM[K]>;
}

export type InputOf<T extends AnyResolver> = T extends Resolver<infer I, any, any, any> ? I : never;

export type OutputOf<T extends AnyResolver> = T extends Resolver<any, infer O, any, any> ? O : never;
