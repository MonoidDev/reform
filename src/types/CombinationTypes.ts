import { isRight, makeLeft } from './Either';
import { ErrorMessage } from './ErrorMessage';
import { AnyResolver, OutputOf } from './infer';
import {
  ErrorContext,
  makeErrorContext,
  makeResolver,
  Resolver,
} from './Resolver';

export interface TUnion<CS extends [AnyResolver, ...AnyResolver[]]>
  extends Resolver<unknown, OutputOf<CS[number]>, ErrorMessage, 'union'> {
  types: CS;
}

export function union<CS extends [AnyResolver, ...AnyResolver[]]>(
  cs: CS,
): TUnion<CS> {
  return {
    ...makeResolver<unknown, OutputOf<CS[number]>, ErrorMessage, 'union'>(
      'union',
    ),
    resolve(input, contexts = [makeErrorContext()]) {
      const newErrorContexts: ErrorContext[] = [];
      let i = 0;
      for (const resolver of cs) {
        const nextContext: ErrorContext = {
          inputPath: [...contexts[contexts.length - 1].inputPath],
          resolverPath: [...contexts[contexts.length - 1].resolverPath, i],
          error: undefined,
        };

        const result = resolver.resolve(input, [nextContext]);

        if (isRight(result)) {
          return result;
        } else {
          newErrorContexts.push(...result.left);
        }
        i++;
      }

      contexts.pop();
      return makeLeft([...contexts, ...newErrorContexts]);
    },
    types: cs,
  };
}
