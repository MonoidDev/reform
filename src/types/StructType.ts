import { isLeft, makeLeft, makeRight } from './Either';
import { ErrorMessage } from './ErrorMessage';
import { InputOfResolverMap, OutputOfResolverMap, ResolverMap } from './infer';
import {
  ErrorContext,
  makeErrorContext,
  makeResolver,
  Resolver,
} from './Resolver';
import { isObject } from './utils';

export interface TStruct<RM extends ResolverMap>
  extends Resolver<
    InputOfResolverMap<RM>,
    OutputOfResolverMap<RM>,
    ErrorMessage,
    'struct'
  > {
  props: RM;
}

export function struct<RM extends ResolverMap>(props: RM): TStruct<RM> {
  return {
    ...makeResolver('struct'),
    resolve(input, contexts = [makeErrorContext()]) {
      if (!isObject(input)) {
        contexts[contexts.length - 1].error = {
          message: `Expect an object, got ${typeof input}`,
        };
        return makeLeft(contexts);
      }

      let value: any = {};

      const newErrorContexts: ErrorContext[] = [];

      for (const [key, resolver] of Object.entries(props)) {
        const nextContext: ErrorContext = {
          inputPath: [...contexts[contexts.length - 1].inputPath, key],
          resolverPath: [...contexts[contexts.length - 1].resolverPath, key],
          error: undefined,
        };
        const result = resolver.resolve(input[key], [nextContext]);

        if (isLeft(result)) {
          newErrorContexts.push(...result.left);
        } else {
          value[key] = result.right;
        }
      }

      if (newErrorContexts.length) {
        contexts.pop(); // Remove last context because there're no errors
        return makeLeft([...contexts, ...newErrorContexts]);
      }

      return makeRight(value);
    },
    props,
  };
}
