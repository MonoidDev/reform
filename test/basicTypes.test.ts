import {
  any,
  array,
  boolean,
  nullType,
  number,
  object,
  string,
  undefinedType,
} from '../src/types/BasicTypes';
import { InputOf, OutputOf } from '../src/types/infer';
import { literal } from '../src/types/LiteralType';
import { struct } from '../src/types/StructType';
import { expectLeft, expectRight } from './utils';

describe('any', () => {
  it('can validate', () => {
    expectRight(any().resolve(1));
    expectRight(any().resolve('1'));
    expectRight(any().resolve({ 1: 1 }));
    expectRight(any().resolve(function () {}));
  });
});

describe('string', () => {
  it('can validate', () => {
    expectRight(string().resolve('114514'));
    expectRight(string().resolve(''));
    expectLeft(string().resolve({}));
    expectLeft(string().resolve(1));
    expectLeft(string().resolve(null));
    expectLeft(string().resolve(undefined));
    expectLeft(string().resolve([]));
  });
});

describe('number', () => {
  it('can validate', () => {
    expectRight(number().resolve(114514));
    expectRight(number().resolve(0));
    expectRight(number().resolve(Infinity));
    expectRight(number().resolve(-1));
    expectRight(number().resolve(NaN));
    expectRight(number().resolve(1.1111));
    expectRight(number().resolve(Number.MAX_VALUE));
    expectRight(number().resolve(Number.MIN_VALUE));

    expectLeft(number().resolve({}));
    expectLeft(number().resolve('1'));
    expectLeft(number().resolve(null));
    expectLeft(number().resolve(undefined));
    expectLeft(number().resolve([]));
  });
});

describe('boolean', () => {
  it('can validate', () => {
    expectRight(boolean().resolve(true));
    expectRight(boolean().resolve(false));

    expectLeft(boolean().resolve({}));
    expectLeft(boolean().resolve('1'));
    expectLeft(boolean().resolve(null));
    expectLeft(boolean().resolve(undefined));
    expectLeft(boolean().resolve([]));
    expectLeft(boolean().resolve(1));
  });
});

describe('null', () => {
  it('can validate', () => {
    expectRight(nullType().resolve(null));

    expectLeft(nullType().resolve({}));
    expectLeft(nullType().resolve('1'));
    expectLeft(nullType().resolve(undefined));
    expectLeft(nullType().resolve([]));
    expectLeft(nullType().resolve(1));
  });
});

describe('undefined', () => {
  it('can validate', () => {
    expectRight(undefinedType().resolve(undefined));

    expectLeft(undefinedType().resolve({}));
    expectLeft(undefinedType().resolve('1'));
    expectLeft(undefinedType().resolve(null));
    expectLeft(undefinedType().resolve([]));
    expectLeft(undefinedType().resolve(1));
  });
});

describe('object', () => {
  it('can validate', () => {
    expectRight(object().resolve({}));
    expectRight(object().resolve(() => {}));
    expectRight(object().resolve(function () {}));
    expectRight(object().resolve([]));

    expectLeft(object().resolve(undefined));
    expectLeft(object().resolve('1'));
    expectLeft(object().resolve(null));
    expectLeft(object().resolve(1));
  });
});

describe('struct', () => {
  const person = struct({
    age: number(),
    name: string(),
  });

  const nested = struct({
    nested: struct({
      age: number(),
    }),
    outer: number(),
  });

  const multi = struct({
    nested: struct({
      nested1: struct({
        age: number(),
      }),
      nested2: struct({
        age: number(),
        name: string(),
      }),
    }),
  });

  const multi2 = struct({
    a: number(),
    b: number(),
    c: number(),
  });

  it('can type', () => {
    const personO: OutputOf<typeof person> = {
      age: 1,
      name: 'Shit',
    };

    const personI: InputOf<typeof person> = {
      age: '1',
      name: {},
    };

    const nestedO: OutputOf<typeof nested> = {
      nested: {
        age: 1,
      },
      outer: 1,
    };
  });

  it('can validate', () => {
    expectRight(
      person.resolve({
        age: 0,
        name: 'shit',
      })
    );

    expectLeft(
      person.resolve({
        age: '1',
        name: NaN,
      })
    );

    expectRight(
      nested.resolve({
        nested: {
          age: 1,
        },
        outer: 123,
      })
    );

    expectLeft(
      nested.resolve({
        nested: {
          age: '1',
        },
        outer: 123,
      })
    );

    expect(
      nested.resolve({
        nested: {
          age: '1',
        },
        outer: 123,
      })
    ).toMatchSnapshot();

    expectLeft(
      multi.resolve({
        nested: {
          nested1: {
            age: '1123',
          },
          nested2: {
            age: '12321',
            name: 123213,
          },
        },
      })
    );

    expect(
      multi.resolve({
        nested: {
          nested1: {
            age: '1123',
          },
          nested2: {
            age: '12321',
            name: 123213,
          },
        },
      })
    ).toMatchSnapshot();

    expectLeft(
      multi.resolve({
        nested: {
          nested1: {
            age: '1123',
          },
          nested2: undefined as any,
        },
      })
    );

    expect(
      multi.resolve({
        nested: {
          nested1: {
            age: '1123',
          },
          nested2: undefined as any,
        },
      })
    ).toMatchSnapshot();
  });
});

describe('array', () => {
  const simple = array(number());

  const objects = array(struct({
    age: number(),
    gender: string(),
  }));

  it('can type', () => {
    const a1: OutputOf<typeof simple> = [1, 2, 3];

    const a2: OutputOf<typeof objects> = [
      {
        age: 1,
        gender: "F"
      },
      {
        age: 2,
        gender: "M",
      },
    ];
  });

  it('can validate', () => {
    expectRight(simple.resolve([1, 2, 3]));
    expectLeft(simple.resolve([1, 2, "3"]));
    expect(simple.resolve([1, 2, "3"])).toMatchSnapshot();

    const inner = struct({
      array: array(string()),
    });

    expectLeft(inner.resolve({
      array: [{}, 1, 'sad'],
    }));

    expect(inner.resolve({
      array: [{}, 1, 'sad'],
    })).toMatchSnapshot();
  });
});

describe('literal', () => {
  const simple = literal('A');

  it('can type', () => {
    const a: OutputOf<typeof simple> = 'A';
  })

  it('can validate', () => {
    expectRight(simple.resolve('A'));
    expectLeft(simple.resolve('B'));
    expect(simple.resolve('B')).toMatchSnapshot();
  });
});
