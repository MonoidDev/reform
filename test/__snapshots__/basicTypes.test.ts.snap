// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`array can validate 1`] = `
Object {
  "_tag": "left",
  "left": Array [
    Object {
      "error": Object {
        "message": "Expect a number, got string. ",
      },
      "inputPath": Array [
        2,
      ],
      "resolverPath": Array [],
    },
  ],
}
`;

exports[`array can validate 2`] = `
Object {
  "_tag": "left",
  "left": Array [
    Object {
      "error": Object {
        "message": "Expect a string, got object. ",
      },
      "inputPath": Array [
        "array",
        0,
      ],
      "resolverPath": Array [
        "array",
      ],
    },
    Object {
      "error": Object {
        "message": "Expect a string, got number. ",
      },
      "inputPath": Array [
        "array",
        1,
      ],
      "resolverPath": Array [
        "array",
      ],
    },
  ],
}
`;

exports[`literal can validate 1`] = `
Object {
  "_tag": "left",
  "left": Array [
    Object {
      "error": Object {
        "message": "Expect A, got B. ",
      },
      "inputPath": Array [],
      "resolverPath": Array [],
    },
  ],
}
`;

exports[`struct can validate 1`] = `
Object {
  "_tag": "left",
  "left": Array [
    Object {
      "error": Object {
        "message": "Expect a number, got string. ",
      },
      "inputPath": Array [
        "nested",
        "age",
      ],
      "resolverPath": Array [
        "nested",
        "age",
      ],
    },
  ],
}
`;

exports[`struct can validate 2`] = `
Object {
  "_tag": "left",
  "left": Array [
    Object {
      "error": Object {
        "message": "Expect a number, got string. ",
      },
      "inputPath": Array [
        "nested",
        "nested1",
        "age",
      ],
      "resolverPath": Array [
        "nested",
        "nested1",
        "age",
      ],
    },
    Object {
      "error": Object {
        "message": "Expect a number, got string. ",
      },
      "inputPath": Array [
        "nested",
        "nested2",
        "age",
      ],
      "resolverPath": Array [
        "nested",
        "nested2",
        "age",
      ],
    },
    Object {
      "error": Object {
        "message": "Expect a string, got number. ",
      },
      "inputPath": Array [
        "nested",
        "nested2",
        "name",
      ],
      "resolverPath": Array [
        "nested",
        "nested2",
        "name",
      ],
    },
  ],
}
`;

exports[`struct can validate 3`] = `
Object {
  "_tag": "left",
  "left": Array [
    Object {
      "error": Object {
        "message": "Expect a number, got string. ",
      },
      "inputPath": Array [
        "nested",
        "nested1",
        "age",
      ],
      "resolverPath": Array [
        "nested",
        "nested1",
        "age",
      ],
    },
    Object {
      "error": Object {
        "message": "Expect an object, got undefined",
      },
      "inputPath": Array [
        "nested",
        "nested2",
      ],
      "resolverPath": Array [
        "nested",
        "nested2",
      ],
    },
  ],
}
`;
